### PR TITLE
Control Monitor app sample

### PR DESCRIPTION
**Pull request by kasumman**
Adding a sample app project that has c++/winrt component to use [IMFCameraControlMonitor]( https://docs.microsoft.com/en-us/windows/win32/api/mfidl/nn-mfidl-imfcameracontrolmonitor)

The sample app listens notifications from two controls, brightness and contrast. 
CameraControlMonitorSampleApp_inaction

### CODE REVIEW

File: Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+using Windows.Devices.Enumeration;
+using Windows.Media.Capture;
+using Windows.Media.Capture.Frames;
+using Windows.Media.Core;
+using Windows.Media.Playback;
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Controls.Primitives;
+using Windows.UI.Xaml.Navigation;
+// The Blank Page item template is documented at https://go.microsoft.com/fwlink/?LinkId=402352&clcid=0x409
```
Comment (LPBourret) [inline Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs:18]: nit: remove

File: Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+            {
+                // This app could define "SharingMode= MediaCaptureSharingMode.SharedReadOnly" here if
+                // the app only wants to be informed of the changes and not have full control camera app.
+                VideoDeviceId = m_cameraDeviceList[selectedIndex].Id,
+                StreamingCaptureMode = StreamingCaptureMode.Video
+            };
+
+            await m_mediaCapture.InitializeAsync(initSettings);
+
+
```
Comment (LPBourret) [inline Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs:77]: nit:1 line space too many

File: Samples/ControlMonitorApp/ControlMonitorHelper/ControlMonitorHelper.idl [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    };
+
+    [default_interface]
+    runtimeclass ControlMonitorManager
+    {
+        static ControlMonitorManager CreateCameraControlMonitor(String symlink, ControlData[] controls);
+        ControlMonitorManager(String cameraSymbolicLink, ControlData[] controls);
+        void Stop();
+        void Start();
+        event Windows.Foundation.EventHandler<ControlData> VidCapCameraControlChanged;
```
Comment (LPBourret) [inline Samples/ControlMonitorApp/ControlMonitorHelper/ControlMonitorHelper.idl:29]: rename if you are going to return callback that specifies one of the 3 enum value in ControlKind

File: Samples/ControlMonitorApp/ControlMonitorHelper/ControlMonitorHelper.cpp [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+    ControlMonitorManager::~ControlMonitorManager() noexcept try
+    {
+        if (m_spMonitor != nullptr)
+        {
+            THROW_IF_FAILED(m_spMonitor->Stop());
+            for each (auto && var in m_controls)
+            {
+                GUID guid{};
+                switch (var.controlKind)
```
Comment (LPBourret) [inline Samples/ControlMonitorApp/ControlMonitorHelper/ControlMonitorHelper.cpp:69]: hmm.. why is this there and why is line 84 just using hardcoded definition of what to remove subscription for?

File: Samples/ControlMonitorApp/ControlMonitorHelper/ControlMonitorHelper.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            }
+
+            //
+            // IMFCameraControlNotify
+            IFACEMETHODIMP_(void) OnChange(_In_ REFGUID controlSet, _In_ UINT32 id) override;
+            IFACEMETHODIMP_(void) OnError(_In_ HRESULT hrStatus) override;
+
+            LONG m_lRef = 0;
+            wil::critical_section m_lock;
+            winrt::event<Windows::Foundation::EventHandler<ControlData>> m_vidCapCameraControlChangedEvent;
```
Comment (LPBourret) [inline Samples/ControlMonitorApp/ControlMonitorHelper/ControlMonitorHelper.h:37]: rename as you could add a callback for other type of controls than vidcap

File: Samples/ControlMonitorApp/ControlMonitorHelper/ControlMonitorHelper.vcxproj [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    <ProjectGuid>{3ac253a5-a383-49d8-acc0-102c1e46e42f}</ProjectGuid>
+    <ProjectName>ControlMonitorHelper_WinRT</ProjectName>
+    <RootNamespace>ControlMonitorHelper</RootNamespace>
+    <DefaultLanguage>en-US</DefaultLanguage>
+    <MinimumVisualStudioVersion>14.0</MinimumVisualStudioVersion>
+    <AppContainerApplication>true</AppContainerApplication>
+    <ApplicationType>Windows Store</ApplicationType>
+    <ApplicationTypeRevision>10.0</ApplicationTypeRevision>
+    <WindowsTargetPlatformVersion Condition=" '$(WindowsTargetPlatformVersion)' == '' ">10.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformMinVersion>10.0.22570.0</WindowsTargetPlatformMinVersion>
```
Comment (LPBourret) [inline Samples/ControlMonitorApp/ControlMonitorHelper/ControlMonitorHelper.vcxproj:18]: is this actually 22000 or 22621?

**Review by LPBourret — CHANGES_REQUESTED** [review]

**Comment by michaelajax** [comment]
Do you mind adding a README.md describing what this does, who would be interested in it, and including a photo/GIF of it in action?

**Comment by LPBourret** [comment]
> 

also add a link to this sample from the readme.md on landing page of this repo please

File: Samples/ControlMonitorApp/CameraControlMonitorAppSample/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("CameraControlMonitorAppSample")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("CameraControlMonitorAppSample")]
+[assembly: AssemblyCopyright("Copyright ©  2021")]
```
Comment (NaveenThum) [inline Samples/ControlMonitorApp/CameraControlMonitorAppSample/Properties/AssemblyInfo.cs:13]: >2021 [](http://example.com/codeflow?start=43&length=4)

2022?

File: Samples/ControlMonitorApp/CameraControlMonitorAppSample/CameraControlMonitorAppSample.csproj [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    <AppxAutoIncrementPackageRevision>True</AppxAutoIncrementPackageRevision>
+    <GenerateTestArtifacts>True</GenerateTestArtifacts>
+    <AppxBundle>Always</AppxBundle>
+    <AppxBundlePlatforms>x64</AppxBundlePlatforms>
+    <HoursBetweenUpdateChecks>0</HoursBetweenUpdateChecks>
+    <PackageCertificateThumbprint>
+    </PackageCertificateThumbprint>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x86'">
+    <DebugSymbols>true</DebugSymbols>
```
Comment (NaveenThum) [inline Samples/ControlMonitorApp/CameraControlMonitorAppSample/CameraControlMonitorAppSample.csproj:33]: Recommend dropping 32-bit apps.

File: Samples/ControlMonitorApp/CameraControlMonitorAppSample.sln [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "ControlMonitorHelper_WinRT", "ControlMonitorHelper\ControlMonitorHelper.vcxproj", "{3AC253A5-A383-49D8-ACC0-102C1E46E42F}"
+EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "CameraKsPropertyHelper", "..\Shared\CameraKsPropertyHelper\CameraKsPropertyHelper.vcxproj", "{3E174357-0294-40F1-BA11-7330617DDBB5}"
+EndProject
+Global
+	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|ARM = Debug|ARM
+		Debug|ARM64 = Debug|ARM64
+		Debug|x64 = Debug|x64
```
Comment (NaveenThum) [inline Samples/ControlMonitorApp/CameraControlMonitorAppSample.sln:16]: Recommend dropping 32-bit apps.

File: Samples/ControlMonitorApp/ControlMonitorHelper/ControlMonitorHelper.cpp (hunk 4 above)
Comment (kasumman) [reply Samples/ControlMonitorApp/ControlMonitorHelper/ControlMonitorHelper.cpp:69]: Good point, missed this while modifying the codebase from Michael to support more than just one hard coded control.

**Review by NaveenThum — APPROVED** [review]
:shipit:

File: Samples/ControlMonitorApp/readme.md [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+## Requirements
+This sample is built using Visual Studio 2022 and requires [Windows SDK version 22621](https://developer.microsoft.com/en-us/windows/downloads/windows-sdk/).
+
+## Sample functionality
+The sample will enumerate available cameras and initialize and start the preview for one of them, user can change the used camera from the combo box.
+The app will then create control listeners for brightness and contrast controls (assuming these are supported by the selected camera) and will update its slider values if an external application changes these values for the same camera.
+
+The Camera Settings page can be used as the external application to demonstrate the control change events in the sample app. I.e., when the contrast slider is changed from the Camera Settings page, the sample application will reflect this change.
+**NOTE:** Moving the sliders in the sample app **will not** be reflected in the Camera Settings sliders as those sliders are reflecting default values that are loaded unless an application overrides the values in its session.
```
Comment (michaelajax) [inline Samples/ControlMonitorApp/readme.md:13]: Recommend elaborating a bit on why this one way behavior exists. Here's a proposal:

> **NOTE:** Moving the sliders in the sample app adjusts the _current_ value of the controls on the camera. The Windows Camera Settings app always displays the _default_ value. When you adjust a setting in the Windows Camera Settings app, the new default value is saved for the user, and the current value on the camera is also set to match. When you adjust a slider in the sample app, the current value on the camera is changed, but the value displayed in the Windows Camera Settings app **will not** be updated.

File: Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+            UIMediaPlayerElement.SetMediaPlayer(m_mediaPlayer);
+
+            // Creating a controls data for listening changes on these controls.
+            var controlContrast = new ControlMonitorHelper.ControlData()
+            {
+                controlKind = ControlMonitorHelper.ControlKind.VidCapVideoProcAmpKind,
+                controlId = (UInt32)CameraKsPropertyHelper.VidCapVideoProcAmpKind.KSPROPERTY_VIDEOPROCAMP_CONTRAST
+            };
+
+            var controlBrightness = new ControlMonitorHelper.ControlData()
```
Comment (michaelajax) [inline Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs:106]: It looks like we are always writing the Brightness control without considering the EV Compensation control.

So, on a camera that supports EV Compensation and Brightness, the app would be changing a different control than the Settings app is changing, and the sync won't work.

Can we update the logic here to match what the Settings app does? Basically it is:

```
if (Camera supports EV Compensation)
   Brightness slider controls EV Compensation
elif (Camrea supports Brightness)
   Brightness slider controls Brightness
else
   Brightness slider not shown
```

**Review by michaelajax — CHANGES_REQUESTED** [review]

File: Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs (hunk 11 above)
Comment (kasumman) [reply Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs]: added

File: Samples/ControlMonitorApp/readme.md [hunk 12]
```diff
@@ -7,9 +7,11 @@ This sample is built using Visual Studio 2022 and requires [Windows SDK version
 
 ## Sample functionality
 The sample will enumerate available cameras and initialize and start the preview for one of them, user can change the used camera from the combo box.
-The app will then create control listeners for brightness and contrast controls (assuming these are supported by the selected camera) and will update its slider values if an external application changes these values for the same camera.
+The app will then create control listeners for brightness[^1] and contrast controls (assuming these are supported by the selected camera) and will update its slider values if an external application changes these values for the same camera.
+
+[^1]: Brightness is replaces with ExposureCompensation control if camera supports it. This is done to have similar functionality as the Camera Settings page.
```
Comment (michaelajax) [inline Samples/ControlMonitorApp/readme.md:12]: nit: replaced instead of replaces

**Review by michaelajax — COMMENTED** [review]
Thanks for the changes! Looks great - one small nit in the readme.

**Review by michaelajax — APPROVED** [review]

File: Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs [hunk 13]
```diff
@@ -124,12 +136,24 @@ private async Task InitializeAsync(int selectedIndex)
...
+                    BrightnessSlider.Visibility = Visibility.Visible;
+                    BrightnessSlider.Value = m_mediaCapture.VideoDeviceController.ExposureCompensationControl.Value;
+                }
+                else
+                {
+                    BrightnessSlider.Minimum = m_mediaCapture.VideoDeviceController.Brightness.Capabilities.Min;
+                    BrightnessSlider.Maximum = m_mediaCapture.VideoDeviceController.Brightness.Capabilities.Max;
+                    BrightnessSlider.StepFrequency = m_mediaCapture.VideoDeviceController.Brightness.Capabilities.Step;
+                    BrightnessSlider.Visibility = Visibility.Visible;
+                    m_mediaCapture.VideoDeviceController.Brightness.TryGetValue(out value);
```
Comment (LPBourret) [inline Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs:154]: nit: handle boolean result returned

File: Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs [hunk 14]
```diff
@@ -192,9 +224,17 @@ private void ContrastSlider_ValueChanged(object sender, RangeBaseValueChangedEve
             m_mediaCapture.VideoDeviceController.Contrast.TrySetValue(e.NewValue);
```
Comment (LPBourret) [inline Samples/ControlMonitorApp/CameraControlMonitorAppSample/MainPage.xaml.cs:224]: nit: same here, handle boolean return result

**Review by LPBourret — APPROVED** [review]

**Pull request merged**